notebooks/circuits_with_bloqade.py

```diff
@@ -0,0 +1,763 @@
+    nshots - The number of times to run the task
+    return_state - If True, also return the density matrix of the qubits after all runs
+    """
+    rho_all:list[np.linalg._linalg.EighResult] = []
+    results = []
+    for _ in range(nshots):
+        result = task.run()
+        results.append(result)
+        if return_state:
+            rho_all.append(PyQrackSimulatorBase.quantum_state(task.qubits()))
+    
+    # Convert ilist to tuple so that it is hashable by Counter https://github.com/QuEraComputing/bloqade-circuit/issues/428
+    def convert(data):
+        if isinstance(data, (list,IList)):
+            return tuple(convert(item) for item in data)
+        return data
+    results = convert(results)
+    
+    data = {key: value/len(results) for key, value in Counter(results).items()} # Normalize to probabilities
+    
+    if return_state:
+        state = np.linalg._linalg.EighResult(eigenvectors = np.concatenate([state.eigenvectors for state in rho_all], axis=1),
+                            eigenvalues = np.concatenate([state.eigenvalues for state in rho_all], axis=0)/len(rho_all))
+        state = canonicalize(state)
+        return data, state
+    else:
+        return data
+
+
+def multistate(task:PyQrackSimulatorTask,nshots:int)-> np.linalg._linalg.EighResult:
+    """
+    Collect the state of the qubits returned by a kernel run by a task.
+    This is similar to multirun, but assumes that the signature of the task's kernel
+    returns a list of qubits.
+    
+    task - A PyQrackSimulatorTask object created from a kernel, whose return signature is a list of qubits
+    nshots - The number of times to run the task
+    """
+    states = []
+    for _ in range(nshots):
+        qbs = task.run()
+        states.append(emulator.quantum_state(qbs))
+        
+    state = np.linalg._linalg.EighResult(eigenvectors = np.concatenate([state.eigenvectors for state in states], axis=1),
+                            eigenvalues = np.concatenate([state.eigenvalues for state in states], axis=0)/len(states))
+    
+    return canonicalize(state)
+
+
+# %%
+# Define a kernel that randomly measures 1 or 0 with 50% probability each.
+try:
+    del coinflip
+except:
+    pass
+@squin.kernel
+def coinflip()-> list[bool]:
+    qubit = squin.qubit.new(2)
+    squin.gate.h(qubit[0])
+    return squin.qubit.measure(qubit)
+
+# Define the emulator and task
+emulator = StackMemorySimulator()
+task = emulator.task(coinflip)
+results, rho = multirun(task, nshots=1000, return_state=True)
+print("Results:", results)
+print("State:  ", rho)
+
+# %% [markdown]
+# # Composition of kernels
+# 
+# The powerful thing about bloqade kernels is that they allow all of the typical syntax of for loops, if-else statements, function calls, and other powerful abstractions. Lets use this to write efficient representations of complex circuits.
+# 
+# For this example, we will use a Trotterization of the 1d Transverse Ising model.
```

Comment:
This single sentence has a lot of details. In fact, you could probably fill an entire tutorial with it 😄 
Please add some explanations here (and some references). Specifically:

* What is trotterization?
* Definition (and some explanation) of the 1d transverse Ising model.
* How are you implementing this in the following? You briefly mention details of composing kernels, but not precisely how you will use them here.

Reply:
Actually, now that I thought a bit more about this, why not make this a separate tutorial? This one is quite long already, so it might be worth splitting some parts out.